third_party/move/move-vm/runtime/src/session.rs

```diff
@@ -523,16 +523,28 @@ impl<'r, 'l> Session<'r, 'l> {
         self.move_vm.runtime.loader().ty_builder()
     }
 
-    pub fn fetch_struct_ty_by_idx(
+    /// If type is a (generic or non-generic) struct or enum, returns its name. Otherwise, returns
+    /// [None].
+    pub fn get_struct_name(
         &self,
-        idx: StructNameIndex,
-        module_storage: &impl ModuleStorage,
-    ) -> Option<Arc<StructType>> {
-        self.move_vm
-            .runtime
-            .loader()
-            .fetch_struct_ty_by_idx(idx, &self.module_store, module_storage)
-            .ok()
+        ty: &Type,
+        module_storage: &dyn ModuleStorage,
+    ) -> PartialVMResult<Option<(ModuleId, Identifier)>> {
+        use Type::*;
+
+        Ok(match ty {
+            Struct { idx, .. } | StructInstantiation { idx, .. } => {
+                let struct_identifier = self
+                    .move_vm
+                    .runtime
+                    .loader()
+                    .struct_name_index_map(module_storage)
```

Comment:
what if the module contains the struct is not loaded?

Reply:
The module cannot contain non-loaded struct? Or what do you mean?

My goal here is to do the following:
1. If we have a struct `Type`, it contains an index.
2. The only way to create the index is via `struct_name_index_map` (I guess having a public inner field + constructor is a problem you are referring to? It might make sense to enforce that we can construct indices only via the map, or it should be test-only construction)
3. Hence, we must get the correct indexed name.

Reply:
@zekun000 updated the code so that struct name index has private constructor and only can be got via indexed map.

Reply:
It also assumes that the map can never be cleared or things removed, right?

Reply:
Yes, but that is also true for other caches? We stored indices in `Type::Struct { .. }`s, and these types can be in per-frame cache, etc. cached by interpreter. If struct name index map is flashed during the transaction execution, we have dangling indices.